### DISCUSSION TITLE
Is "safari10" "mangle" option implementation even necessary?

### DISCUSSION THREAD

**Discussion by joebeachjoebeach** [swc-project/swc#3373] [Q&A] [answered]
Hello!

I see from [the `minify.mangle` docs](https://swc.rs/docs/configuration/minification#jscminifymangle) that the `safari10` option (available in `terser`) has not yet been implemented.

That terser option is meant to prevent code like this from being generated:
```js
function loop() {
    let a = someFunc();
    console.log(a);
    for (let a of [1,2,3]) {
        console.log(a);
    }
}
```

[Safari 10 views the second `let a` (in the `for` loop) as a syntax error](https://bugs.webkit.org/show_bug.cgi?id=171041).

---

I took a look at what terser and swc do with this code.

#### `safari10: false`

```js
function o(){let o=someFunc();console.log(o);for(let o of[1,2,3])console.log(o)}
```

Duplicate variable name: `o` is declared twice. This will be a syntax error in Safari 10.

#### `safari10: true`

```js
function o(){let o=someFunc();console.log(o);for(let l of[1,2,3])console.log(l)}
```

No more duplicate variable name. The second variable gets a new name: `l`. This will not be a syntax error in Safari 10.

#### `safari10: false` (the only option currently)

```js
function a(){let a=someFunc();console.log(a);for(let b of[1,2,3]){console.log(b)}}
```

No more duplicate variable name. The first one is `a` and the second is now `b`. This will not be a syntax error in Safari 10.

---

Given this, is implementing the `minify.mangle.safari10` option necessary? Our team is keen on swapping out `terser` for `swc` in our build process, and this makes me feel like we don't really need to wait for `safari10` to specifically be implemented. Am I right in thinking that the bug is avoided by default?

**Comment by kdy1** [comment, marked as answer]
Yes, because `serde` does not provide a way to ignore a field.
It's not used, but it's api of `terser`.

**Reply by kdy1** [reply]
You are right, it will not cause an issue.
But removing may create an issue because of serde issue.

**Reply by joebeachjoebeach** [reply]
Thanks for the quick response!

**Comment by taozhou-glean** [comment]
looks like safari10 is now supported:
<img width="539" alt="image" src="https://github.com/swc-project/swc/assets/121891361/a427d6cf-a11d-48c8-8c51-6abfa1744234">

and it seems to have some side effects once enabled:

(the `s` was redefined within the `catch` block and also in the `catch(s)` )

https://play.swc.rs/?version=1.3.104&code=H4sIAAAAAAAAA8VTTY%2FaMBC98ytG2wtIBDdAYaEf0nahWqRdiugeelvZzoS4JDb1OLC02v9eB1q%2BEqnqqXNI%2FPHe5M17CmMgTYTg%2BBI1xNZkkDi3oiFjC%2BWSXLSkydjNj9zizYhlSlpDJnYBz12C2inJnTI6SJWw3G6D2NjgGzGRGsHkoBOG0SDsD7qiH3dFh2PYFtft%2FpuuCNudXr%2FbwxDFNfNklhFPA2HNhtAyspL59dIvte%2B%2Fxo%2F7zXS3eUAivsA7rqMUbcvRq%2Fvwda%2BGzytjHciUE0EVEn7WwBc5fyeB01ZLkBa5w5k1axWhre8ARaVmsUA7hPvdu3k4T3wrSrxXjypDk7sHGoLOM3GKWaH1NmRcS7xNlTdpCJPZ5dkO3RiC%2F3SmCN9VCf7wW%2FFRUMtZLrF%2BVTlecDENSJ6mGLWuGm8PbZzdnjQtShpNDlYWY7QWowP5PWjcVPpYP%2BMfxTVL52W3ypiSW2XIXsRtIZRrR63ZfPxpPJ%2BPR0%2Fjr4%2Fj6ZfJ5%2BnTZHRGOxm5KL7hqmLIFqGO7v6InOP3HMnVL7gWXW51mXxEvXinnUygjo0LcxmDSXykAj471OR%2FGIi5Sgn85LBJfGZr77gPj4xu%2Bqs0FVwuwRnISekFcL0FVUxf5HlsUpFjwctX%2Fz%2FEv4dxrvRfkjhnnsRQ2z9ffgH%2F94Mb1AQAAA%3D%3D&config=H4sIAAAAAAAAA41V22rcQAx9z1cYQ9%2BW0iRQSh%2F6VEopBPoHRuvReCcZj4aR7OwS8u%2BVb5u9jJfC4rV1NLoead7uiqJ85rr8Xrzpq37gXjAF8L%2FRR0ysgKQONxMYITGmo7JK%2BBAE9iop5RCR6%2BSilJsFFd7PBkbJ%2B2xGEgS2lNpTSx4bqA8%2FsaYEQunMscJmAZ5QwIBAzrAnYjwPuXXB2cOpoxZC4%2FFEMkRE0WOPXqUWPB%2FdKvSCGKvaA3OAFnlNw95EY3I9COo%2FxZySw28ZKYOF5O6%2FnKZ6TFbxmtqYkPk8F0hN12KQnB9IiV75orYq3xJ51KasIxVw5YJgM5Hi0nBN3kNkrHpIGStDpJoKU87FAHaC5lidKzwYJ46UlRnUIJiqJoMZyCWsxfWYO6a%2B9FhgGrlwmc8IG9x2TYOXXBympAffaUMzyH5siUabsbojx1LZLuRKOIErNZjAubhX7LFVQulSuD73TC6s9OR%2Fia18Wjt9m%2FQuWB0%2BOWRwHdRclkF3gE6JczZT2aEymMTlupnQdDUOla1z4czwSvnYGazQWuVKxjS%2FOql3OafDxiObAbS%2FYHOsmoDqOIUr%2BDAQN%2BBfmqXkCTZrtCC7dZQP7Zb8DQctyo7MDQVthdA6nHRL7OM63gWDSg00WZWOR%2BB6CegACFUeJccNHQ%2B1WDWeth9rYtmYd8tz3J0lhv7jwmun1VF2DA3OV5d6S%2FhnsFI%2Bfn7U3yLnnYsRzV%2BlIvHlOipFRwVHDpU6VlI8FL1uS%2B3Upvjx8GlTBJJiWFfTm8Pi%2Fn55rVra6ufiR%2FeOdfsz0k2ha7TdyeU1cnDwh%2Fy1PCo5flrUltMf9%2BBc8Pd%2F5hMUQPsHAAA%3D
